Foundation/NSDecimal.swift

```diff
@@ -1,155 +1,1838 @@
+    for i in (0..<d._length).reversed() {
+        let accumulator = UInt32(d[i]) + carry * (1<<16)
+        d[i] = UInt16(accumulator / UInt32(divisor))
+        carry = accumulator % UInt32(divisor)
+    }
+    while d._length != 0 && d[d._length - 1] == 0 {
+        d._length -= 1
+    }
+    return (UInt16(carry),.noError)
+}
 
+fileprivate func multiplyByShort<T:VariableLengthNumber>(_ d: inout T, _ mul:UInt16) -> NSDecimalNumber.CalculationError {
+    if mul == 0 {
+        d._length = 0
+        return .noError
+    }
+    var carry: UInt32 = 0
+    // FIXME handle NSCalculationOverflow here?
+    for i in 0..<d._length {
+        let accumulator: UInt32 = UInt32(d[i]) * UInt32(mul) + carry
+        carry = accumulator >> 16
+        d[i] = UInt16(truncatingBitPattern: accumulator)
+    }
+    if carry != 0 {
+        if d._length >= Decimal.maxSize {
+            return .overflow
+        }
+        d[d._length] = UInt16(truncatingBitPattern: carry)
+        d._length += 1
+    }
+    return .noError
+}
+
+fileprivate func addShort<T:VariableLengthNumber>(_ d: inout T, _ add:UInt16) -> NSDecimalNumber.CalculationError {
+    var carry:UInt32 = UInt32(add)
+    for i in 0..<d._length {
+        let accumulator: UInt32 = UInt32(d[i]) + carry
+        carry = accumulator >> 16
+        d[i] = UInt16(truncatingBitPattern: accumulator)
+    }
+    if carry != 0 {
+        if d._length >= Decimal.maxSize {
+            return .overflow
+        }
+        d[d._length] = UInt16(truncatingBitPattern: carry)
+        d._length += 1
+    }
+    return .noError
+}
+
+public func NSDecimalIsNotANumber(_ dcm: UnsafePointer<Decimal>) -> Bool {
+	return dcm.pointee.isNaN
+}
 
 /***************	Operations		***********/
-public func NSDecimalCopy(_ destination: UnsafeMutablePointer<Decimal>, _ source: UnsafePointer<Decimal>) { NSUnimplemented() }
+public func NSDecimalCopy(_ destination: UnsafeMutablePointer<Decimal>, _ source: UnsafePointer<Decimal>) {
+    destination.pointee.__lengthAndFlags = source.pointee.__lengthAndFlags
+    destination.pointee.__exponent = source.pointee.__exponent
+    destination.pointee.__reserved = source.pointee.__reserved
+    destination.pointee._mantissa = source.pointee._mantissa
+}
 
-public func NSDecimalCompact(_ number: UnsafeMutablePointer<Decimal>) { NSUnimplemented() }
+public func NSDecimalCompact(_ number: UnsafeMutablePointer<Decimal>) {
+    number.pointee.compact()
+}
 
-public func NSDecimalCompare(_ leftOperand: UnsafePointer<Decimal>, _ rightOperand: UnsafePointer<Decimal>) -> ComparisonResult { NSUnimplemented() }
 // NSDecimalCompare:Compares leftOperand and rightOperand.
+public func NSDecimalCompare(_ leftOperand: UnsafePointer<Decimal>, _ rightOperand: UnsafePointer<Decimal>) -> ComparisonResult {
+    let left = leftOperand.pointee
+    let right = rightOperand.pointee
+    return left.compare(to: right)
+}
+
+fileprivate extension UInt16 {
+    func compareTo(_ other: UInt16) -> ComparisonResult {
```

Comment:
It looks like it's used only once and can be trivially inlined. Doing so would have performance benefit and improve clarity at the use site.